README.md
# pypercrypt 📄🔐

![PyPI - Version](https://img.shields.io/pypi/v/pypercrypt)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/pypercrypt)
[![Checked with mypy](https://img.shields.io/badge/mypy-checked-green)](https://mypy-lang.org/)
[![pre-commit enabled](https://img.shields.io/badge/pre--commit-enabled-success?logo=pre-commit)](./.pre-commit-config.yaml)
[![License EUPL](https://img.shields.io/badge/license-EUPL-green)](./LICENSE)
[![Linting: Ruff](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/charliermarsh/ruff/main/assets/badge/v2.json)](https://github.com/astral-sh/ruff)

`pypercrypt` uses battle-tested cryptography to encrypt your data with the passphrase of your choice
and stores the ciphertext inside a QR code.

```bash
echo "Hello World" > keep-secret.txt
pypercrypt encrypt --input-file keep-secret.txt --output-file ciphertext.json
# Enter passphrase: password
cat ciphertext.json
# {"ciphertext": "gAAAAABlQCb4fKNgX3xI1LHowFUjjXhOf-83DeUgukyUAYAF-hMnQ9eFfNl_V4NBoUuu4ZfAAIkVxIM7Iq2Pfh5JaVW_E-rqXA==", "salt": "JAVWUmK_PSUbvEghki3nNQ=="}
cat ciphertext.json | qrencode -o ciphertext.png
# In future versions pypercrypt will support writing to and reading from QR codes directly
pypercrypt decrypt --input-file <(zbarimg -q --raw ciphertext.png) --output-file -
# Enter passphrase: password
# Hello World
```

![Ciphertext encoded as a QR code](./assets/ciphertext.png)

You can store the printed QR code wherever you would trust other encrypted data to stay safe.

# Benefits

- 🔐 Makes encrypted long-term storage easy
- 📵 Truly offline backups
- 🖨️ Use what you already have
- 🔓 No lock-in, no proprietary formats—you don't need this tool to decrypt your data!

# Peer-Review

A peer review is pending, so keep that in mind when using `pypercrypt` for truly sensitive data.
Get in touch if you have the required expertise to review this tool.

# Installation

## pipx (Recommended)

`pipx` creates a separate virtual environment for each CLI tool it installs and makes the scripts
from the package available in your PATH. See
[installation instructions for `pipx`](https://pypa.github.io/pipx/installation/).

```bash
pipx install pypercrypt
```

## Other

You may install `pypercrypt` with any tool that can install from pypi.org.
[See `pypercrypt` on pypi.org](https://pypi.org/project/pypercrypt/).

# Roadmap

## Now

- [x] Encrypt input data and write ciphertext to file
- [ ] Add option to output ciphertext as a printable QR code
- [x] Decrypt the data, relying on an external QR code scanner like `zbarimg`
- [ ] Reach 100 % test coverage of the critical code paths before adding new features

## Next

- [ ] Decrypt the data straight from an image file containing a QR code
- [ ] Instructions on how to decrypt the data purely with standard tools
- [ ] Maintain 100 % test-coverage of the critical code paths
- [ ] Release v1.0 (until then the CLI and format may be subject to breaking changes)

## Later

- Let relevant experts perform an audit of this software
- Keep `pypercrypt` small and auditable by anyone
- A small scope is a feature
- Resist temptation to add more features

## Never

The following features will never be added to `pypercrypt`, either because they go against its goals or would
make the scope too large.

- Do not generate passphrases for the user
- Do not add asymmetric encryption support
- Do not add a GUI or TUI (if I can't resist, then do it as a separate project using `pypercrypt` as a library)

# Alternatives

- [`papercrypt`](https://github.com/TMUniversal/papercrypt) fulfills a similar purpose, is written in Go and
  not affiliated with this project

# License

Licensed under the EUPL. See [LICENSE](./LICENSE).


CHANGELOG.md
# v0.3.1

- Improve documentation and package properties

# v0.3.0

- Release to PyPi

# v0.2.0

- Implement decryption
- Add reading from files and stdin
- Add writing to files and stdout

# v0.1.0

- Implement encryption


pypercrypt/main.py
#!/usr/bin/env python3
# Copyright (C) 2023 Tom Wolfskämpf. Licensed under the EUPL.
"""
The CLI interface for `pypercrypt`.

`pypercrypt` uses battle-tested cryptography to encrypt your data with the passphrase
of your choice and stores the ciphertext inside a QR code.
"""

import base64
import io
import json
import os
from typing import Annotated

import typer
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

app = typer.Typer()


def turn_passphrase_into_key(
    passphrase: str,
    salt: bytes | None = None,
) -> tuple[bytes, bytes]:
    """Derive key from passphrase.

    May use provided salt.

    Return key and salt.
    """
    if not salt:
        salt = base64.urlsafe_b64encode(os.urandom(16))

    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=480000,
    )

    encoded_passphrase = passphrase.encode("utf-8")

    key = base64.urlsafe_b64encode(kdf.derive(encoded_passphrase))

    return key, salt


def read_cleartext_from_file(file: io.BufferedReader) -> bytes:
    """Read cleartext from a file-like object."""
    return file.read()


def read_ciphertext_from_file(file: io.TextIOWrapper) -> tuple[str, bytes]:
    """Read ciphertext and salt from a file-like object."""
    ciphertext_dict = json.loads(file.read())

    ciphertext = ciphertext_dict.get("ciphertext")
    salt = ciphertext_dict.get("salt").encode("utf-8")

    return ciphertext, salt


def encrypt_fernet(cleartext: bytes, key: str | bytes) -> bytes:
    """Encrypt cleartext with Fernet using the provided key."""
    f = Fernet(key)
    token = f.encrypt(cleartext)
    return token


def decrypt_fernet(ciphertext: str, key: bytes) -> bytes:
    """Decrypt ciphertext with Fernet using the provided key."""
    f = Fernet(key)
    cleartext = f.decrypt(ciphertext)
    return cleartext


@app.command()
def encrypt(
    input_file: Annotated[typer.FileBinaryRead, typer.Option()],
    output_file: Annotated[typer.FileTextWrite, typer.Option()],
) -> None:
    """Encrypt input-file content and write ciphertext to output-file."""
    cleartext = read_cleartext_from_file(input_file)
    passphrase = input("Enter passphrase: ")
    key, salt = turn_passphrase_into_key(passphrase)
    ciphertext = encrypt_fernet(cleartext=cleartext, key=key)
    result_dict = {"ciphertext": ciphertext.decode(), "salt": salt.decode()}
    output_file.write(json.dumps(result_dict))


@app.command()
def decrypt(
    input_file: Annotated[typer.FileText, typer.Option()],
    output_file: Annotated[typer.FileBinaryWrite, typer.Option()],
) -> None:
    """Decrypt input-file content and write cleartext to output-file."""
    ciphertext, salt = read_ciphertext_from_file(input_file)
    passphrase = input("Enter passphrase: ")
    key, _ = turn_passphrase_into_key(passphrase, salt)
    cleartext = decrypt_fernet(ciphertext=ciphertext, key=key)
    output_file.write(cleartext)


src/pypercrypt.py
#!/usr/bin/env python3
# Copyright (C) 2023 Tom Wolfskämpf. Licensed under the EUPL.
"""
The CLI interface for `pypercrypt`.

`pypercrypt` uses battle-tested cryptography to encrypt your data with the passphrase
of your choice and stores the ciphertext inside a QR code.
"""

import base64
import io
import json
import os
from typing import Annotated

import typer
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

app = typer.Typer()


def turn_passphrase_into_key(
    passphrase: str,
    salt: bytes | None = None,
) -> tuple[bytes, bytes]:
    """Derive key from passphrase.

    May use provided salt.

    Return key and salt.
    """
    if not salt:
        salt = base64.urlsafe_b64encode(os.urandom(16))

    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=480000,
    )

    encoded_passphrase = passphrase.encode("utf-8")

    key = base64.urlsafe_b64encode(kdf.derive(encoded_passphrase))

    return key, salt


def read_cleartext_from_file(file: io.BufferedReader) -> bytes:
    """Read cleartext from a file-like object."""
    return file.read()


def read_ciphertext_from_file(file: io.TextIOWrapper) -> tuple[str, bytes]:
    """Read ciphertext and salt from a file-like object."""
    ciphertext_dict = json.loads(file.read())

    ciphertext = ciphertext_dict.get("ciphertext")
    salt = ciphertext_dict.get("salt").encode("utf-8")

    return ciphertext, salt


def encrypt_fernet(cleartext: bytes, key: str | bytes) -> bytes:
    """Encrypt cleartext with Fernet using the provided key."""
    f = Fernet(key)
    token = f.encrypt(cleartext)
    return token


def decrypt_fernet(ciphertext: str, key: bytes) -> bytes:
    """Decrypt ciphertext with Fernet using the provided key."""
    f = Fernet(key)
    cleartext = f.decrypt(ciphertext)
    return cleartext


@app.command()
def encrypt(
    input_file: Annotated[typer.FileBinaryRead, typer.Option()],
    output_file: Annotated[typer.FileTextWrite, typer.Option()],
) -> None:
    """Encrypt input-file content and write ciphertext to output-file."""
    cleartext = read_cleartext_from_file(input_file)
    passphrase = input("Enter passphrase: ")
    key, salt = turn_passphrase_into_key(passphrase)
    ciphertext = encrypt_fernet(cleartext=cleartext, key=key)
    result_dict = {"ciphertext": ciphertext.decode(), "salt": salt.decode()}
    output_file.write(json.dumps(result_dict))


@app.command()
def decrypt(
    input_file: Annotated[typer.FileText, typer.Option()],
    output_file: Annotated[typer.FileBinaryWrite, typer.Option()],
) -> None:
    """Decrypt input-file content and write cleartext to output-file."""
    ciphertext, salt = read_ciphertext_from_file(input_file)
    passphrase = input("Enter passphrase: ")
    key, _ = turn_passphrase_into_key(passphrase, salt)
    cleartext = decrypt_fernet(ciphertext=ciphertext, key=key)
    output_file.write(cleartext)


if __name__ == "__main__":
    app()
